staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/streaming.go

```diff
@@ -0,0 +1,249 @@
+	}
+	typeMetaField, ok := listType.FieldByName("TypeMeta")
+	if !ok {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, fmt.Errorf("expected TypeMeta field")
+	}
+	if typeMetaField.Tag.Get("json") != ",inline" {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, fmt.Errorf("expected TypeMeta field tag")
+	}
+	typeMetaValue := v.FieldByName("TypeMeta")
+	if !typeMetaValue.IsValid() {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, fmt.Errorf("expected TypeMeta field")
+	}
+	typeMeta, ok := typeMetaValue.Interface().(metav1.TypeMeta)
+	if !ok {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, fmt.Errorf("expected TypeMeta field to have TypeMeta type")
+	}
+
+	itemsField, ok := listType.FieldByName("Items")
+	if !ok {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, fmt.Errorf("expected Items field")
+	}
+	if itemsField.Tag.Get("json") != "items" {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, fmt.Errorf("expected TypeMeta field tag")
+	}
+	items, err := meta.ExtractList(list)
+	if err != nil {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, err
+	}
+	return typeMeta, listMeta, items, nil
+}
+
+func streamingEncodeList(w io.Writer, typeMeta metav1.TypeMeta, listMeta metav1.ListMeta, items []runtime.Object) error {
+	_, err := w.Write([]byte(`{`))
+	if err != nil {
+		return err
+	}
+	if typeMeta.Kind != "" {
+		err := encodeValue(w, []byte(`"kind":`), typeMeta.Kind, []byte(","))
+		if err != nil {
+			return err
+		}
+	}
+	if typeMeta.APIVersion != "" {
+		err := encodeValue(w, []byte(`"apiVersion":`), typeMeta.APIVersion, []byte(","))
+		if err != nil {
+			return err
+		}
+	}
+	err = encodeValue(w, []byte(`"metadata":`), listMeta, []byte(`,"items":`))
+	if err != nil {
+		return err
+	}
+	if items == nil {
+		_, err = w.Write([]byte("null}\n"))
+		return err
+	}
+	if len(items) == 0 {
+		_, err = w.Write([]byte("[]}\n"))
+		return err
+	}
+	suffix := []byte(",")
+	prefix := []byte("[")
+	for i, item := range items {
+		if i == len(items)-1 {
+			suffix = nil
+		}
+		err = encodeValue(w, prefix, item, suffix)
+		if err != nil {
+			return err
+		}
+		if i == 0 {
+			prefix = nil
+		}
+	}
+	_, err = w.Write([]byte("]}\n"))
+	return err
+}
+
+func encodeValue(w io.Writer, prefix []byte, value any, suffix []byte) error {
+	data, err := json.Marshal(value)
+	if err != nil {
+		return err
+	}
+	if len(prefix) != 0 {
+		_, err = w.Write(prefix)
+		if err != nil {
+			return err
+		}
+	}
+	_, err = w.Write(data)
+	if err != nil {
+		return err
+	}
+	if len(suffix) != 0 {
+		_, err = w.Write(suffix)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func streamingEncodeUnstructuredList(w io.Writer, list *unstructured.UnstructuredList) error {
+	_, err := w.Write([]byte(`{`))
+	if err != nil {
+		return err
+	}
+	keys := slices.Sorted(maps.Keys(list.Object))
+	index := 0
+	for ; index < len(keys) && keys[index] < "items"; index++ {
```

Comment:
would it be simpler to range over the whole slice, and have a different branch in the loop body when we encounter the "items" key, rather than duplicating the loop with identical code for the before-items and after-items keys?

Reply:
Done